syntax = "proto3";
package raft;

message VoteRequest {
  int32 term = 1;
  uint64 candidateId = 2;
  uint64 lastLogIndex = 3;
  uint64 lastLogTerm = 4;
}

message VoteResponse {
  uint64 nodeId = 1;
  int32 term = 2;
  bool voteGranted = 3;
  double leaderLease = 4;
}

message SuffixEntry {
  string op = 1;
  string key = 2;
  string value = 3;
  int32 term = 4;
}

message LogRequest {
  uint64 leaderId = 1;
  int32 currentTerm = 2;
  uint64 prefixLen = 3;
  int32 prefixTerm = 4;
  uint64 commitLength = 5;
  repeated SuffixEntry suffix = 6;
  double leaderLease = 7;
}

message LogResponse {
  uint64 nodeId = 1;
  int32 term = 2;
  int32 ack = 3;
  bool success = 4;
}

message ServeClientArgs {
  string Request = 1;
}

message ServeClientReply {
  string Data = 1;
  string LeaderID = 2;
  bool Success = 3;
}

service RaftService {
  rpc VoteRequested (VoteRequest) returns (VoteResponse);
  rpc ServeClient (ServeClientArgs) returns (ServeClientReply) {}
  rpc LoggingRequested (LogRequest) returns (LogResponse) {}
  // Other Raft RPCs (e.g., AppendEntries, InstallSnapshot)
}
